### PR TITLE
Fix serialization errors when skipping compilation in `FusionExecutor`. 

### PR DESCRIPTION
**Pull request by rdspring1**
Serialization assumes that `FusionExecutor` always contains a compiled kernel. This PR updates serialization to skip compilation correctly.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: csrc/executor.cpp [hunk 1]
```diff
@@ -2306,6 +2309,17 @@ void FusionExecutor::deserialize(
   // See table definition for FusionExecutor in serde/fusion_cache.fbs
 
   NVF_ERROR(buffer != nullptr, "serde::FusionExecutor is nullptr.");
+
+  // TODO Should we set fusion_id, concrete_id, runtime_id, and group_id when we
+  // skip compilation?
+  if (buffer->is_compilation_skipped()) {
+    fusion_ = std::make_unique<Fusion>(*fusion);
+    NVF_ERROR(
+        !isCompiled() && isCompilationSkipped(),
```
Comment (Priya2698) [inline csrc/executor.cpp:2318]: `!isCompiled()` should be sufficient, why both the conditions?

File: csrc/executor.cpp [hunk 2]
```diff
@@ -2364,7 +2378,9 @@ void FusionExecutor::deserialize(
   compiled_kernel_ = executor_utils::getCompiledKernel(
       buffer->compiled_kernel(), compile_params);
 
-  NVF_ERROR(isCompiled(), "Failed to deserialize FusionExecutor");
+  NVF_ERROR(
+      isCompiled() && !isCompilationSkipped(),
```
Comment (Priya2698) [inline csrc/executor.cpp:2382]: Same as above, do we need both?

**Comment by Priya2698** [comment]
General question: How does nvFuser handle if only one out of the many segments is not compiled? Do we skip serialization/deserialization for the entire fusion or is it segment-wise?

**Comment by rdspring1** [comment]
> General question: How does nvFuser handle if only one out of the many segments is not compiled? Do we skip serialization/deserialization for the entire fusion or is it segment-wise?

We're skipping on a `FusionExecutor` basis, so it is segment-wise. Serialization follows the same behavior.

File: csrc/executor.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/executor.cpp:2318]: The assertion wasn't originally within the if-statement, and I didn't update it.

**Comment by Priya2698** [comment]
There are some new issues on my branch. I prefer either merging this to main or first fixing the issues in my branch so the root cause is clear. Wdyt?

**Comment by Priya2698** [comment]
!build

**Comment by rdspring1** [comment]
This PR can wait until the issues in your branch are fixed.

**Comment by Priya2698** [comment]
> This PR can wait until the issues in your branch are fixed.

@rdspring1 We can merge this into my branch now. Could you resolve the conflicts, and finalize this PR? Thanks for the fix!

**Comment by rdspring1** [comment]
!build

File: csrc/executor.cpp [hunk 3]
```diff
@@ -2178,8 +2178,13 @@ flatbuffers::Offset<serde::FusionExecutor> FusionExecutor::serialize(
     executor_entry_lookup_values_fb.push_back(serialize(builder, value));
   }
 
+  if (isCompilationSkipped()) {
```
Comment (Priya2698) [inline csrc/executor.cpp:2181]: Please add a comment here explaining the if condition

**Review by Priya2698 — APPROVED** [review]
LGTM. Added minor comments.

**Pull request merged**